src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,414 @@
+    for data_selector_id in data_selectors:
+        keys = set(
+            key.partition('.')[0] for key in dc[data_selector_id].keys()
+            if key.endswith(suffix))
+        matches = (re_ptrn_cc.match(src) for src in keys)
+        unique_detectors = set(
+            m["detector_id"] for m in matches if m is not None)
+        if not unique_detectors:
+            matches = (re_ptrn_us.match(src) for src in keys)
+            unique_detectors = set(
+                m["detector_id"] for m in matches if m is not None)
+            unique_detectors = {
+                detector_id: pattern_underscore.format(
+                    detector_id=detector_id, **frm_args)
+                for detector_id in unique_detectors
+            }
+        else:
+            unique_detectors = {
+                guess_device_id(detector_id):
+                pattern_camelcase.format(detector_id=detector_id, **frm_args)
+                for detector_id in unique_detectors
+            }
+
+        for detector_id, src_ptrn in unique_detectors.items():
+            if detector_id in detectors:
+                continue
+
+            num_sources = count_sources(keys, src_ptrn, **dims)
+            if num_sources == num_motors:
+                detectors[detector_id] = (
+                    data_selector_id, src_ptrn + '.' + position_key)
+
+    return detectors
+
+
+def find_motors(dc, pattern, position_key, **dims):
+    """Looks for motors related to given detector in data collection."""
+    num_motors, frm_args, _, placeholders = _make_motor_placeholders(**dims)
+
+    pattern_camelcase = mangle_pattern(
+        pattern, mangle_device_id_camelcase, **placeholders)
+    pattern_underscore = mangle_pattern(
+        pattern, mangle_device_id_underscore, **placeholders)
+
+    src_ptrn = pattern.format(**frm_args)
+    num_sources = count_sources(dc.control_sources, src_ptrn, **dims)
+    if num_sources == num_motors:
+        return src_ptrn, position_key
+
+    data_selectors = []
+    for src in dc.control_sources:
+        try:
+            class_id = dc.get_run_value(src, "classId")
+            if class_id == "SlowDataSelector":
+                data_selectors.append(src)
+        except PropertyNameError:
+            # class id is unknown, skip source
+            pass
+
+    suffix = f".{position_key}.value"
+    for data_selector_id in data_selectors:
+        keys = set(
+            key.partition('.')[0] for key in dc[data_selector_id].keys()
+            if key.endswith(suffix))
+        src_ptrn = pattern_camelcase.format(**frm_args)
+        num_sources = count_sources(keys, src_ptrn, **dims)
+        if num_sources == num_motors:
+            return data_selector_id, src_ptrn + '.' + position_key
+
+        src_ptrn = pattern_underscore.format(**frm_args)
+        num_sources = count_sources(keys, src_ptrn, **dims)
+        if num_sources == num_motors:
+            return data_selector_id, src_ptrn + '.' + position_key
```

Comment:
This is a big chunk of pretty complicated code to identify sources & keys. I guess the aim is to generalise so it will work for other detectors in the future, but I don't think it's worth it - I think writing the code specifically for AGIPD and then doing something similar for other detectors if/when we need it will end up being easier to understand.

Reply:
I agree, it is over complicated. I think I could considerably simplify it, keeping still general. But I have implemented searching only for known or specified detector_id